Library/OcPciIoLib/OcPciIoU.c

```diff
@@ -0,0 +1,446 @@
+//
+UINT8  mOutStride[] = {
+  1,      // EfiCpuIoWidthUint8
+  2,      // EfiCpuIoWidthUint16
+  4,      // EfiCpuIoWidthUint32
+  8,      // EfiCpuIoWidthUint64
+  1,      // EfiCpuIoWidthFifoUint8
+  2,      // EfiCpuIoWidthFifoUint16
+  4,      // EfiCpuIoWidthFifoUint32
+  8,      // EfiCpuIoWidthFifoUint64
+  0,      // EfiCpuIoWidthFillUint8
+  0,      // EfiCpuIoWidthFillUint16
+  0,      // EfiCpuIoWidthFillUint32
+  0       // EfiCpuIoWidthFillUint64
+};
+
+EFI_CPU_IO2_PROTOCOL *
+InitializeCpuIo2 (
+  )
+{
+  mCpuIo = NULL;
+  gBS->LocateProtocol (
+         &gEfiCpuIo2ProtocolGuid,
+         NULL,
+         (VOID **)&mCpuIo
+         );
+  return mCpuIo;
+}
+
+/**
+   Check parameters to a CPU I/O 2 Protocol service request.
+
+   The I/O operations are carried out exactly as requested. The caller is responsible
+   for satisfying any alignment and I/O width restrictions that a PI System on a
+   platform might require. For example on some platforms, width requests of
+   EfiCpuIoWidthUint64 do not work. Misaligned buffers, on the other hand, will
+   be handled by the driver.
+
+   @param[in] MmioOperation  TRUE for an MMIO operation, FALSE for I/O Port operation.
+   @param[in] Width          Signifies the width of the I/O or Memory operation.
+   @param[in] Address        The base address of the I/O operation.
+   @param[in] Count          The number of I/O operations to perform. The number of
+                            bytes moved is Width size * Count, starting at Address.
+   @param[in] Buffer         For read operations, the destination buffer to store the results.
+                            For write operations, the source buffer from which to write data.
+
+   @retval EFI_SUCCESS            The parameters for this request pass the checks.
+   @retval EFI_INVALID_PARAMETER  Width is invalid for this PI system.
+   @retval EFI_INVALID_PARAMETER  Buffer is NULL.
+   @retval EFI_UNSUPPORTED        The Buffer is not aligned for the given Width.
+   @retval EFI_UNSUPPORTED        The address range specified by Address, Width,
+                                 and Count is not valid for this PI system.
+
+ **/
+EFI_STATUS
+CpuIoCheckParameter (
+  IN BOOLEAN                    MmioOperation,
+  IN EFI_CPU_IO_PROTOCOL_WIDTH  Width,
+  IN UINT64                     Address,
+  IN UINTN                      Count,
+  IN VOID                       *Buffer
+  )
+{
+  UINT64  MaxCount;
+  UINT64  Limit;
+
+  //
+  // Check to see if Buffer is NULL
+  //
+  if (Buffer == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  //
+  // Check to see if Width is in the valid range
+  //
+  if ((UINT32)Width >= EfiCpuIoWidthMaximum) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  //
+  // For FIFO type, the target address won't increase during the access,
+  // so treat Count as 1
+  //
+  if ((Width >= EfiCpuIoWidthFifoUint8) && (Width <= EfiCpuIoWidthFifoUint64)) {
+    Count = 1;
```

Comment:
This code overwrites Count, but the caller of this code (e.g. CpuMemoryServiceRead) is unaware of this change. Is this intentional?

Reply:
This code comes from EDK2, so I guess this is within spec.